src/backend/distributed/metadata/metadata_utility.c

```diff
@@ -2357,3 +2365,1558 @@ IsForeignTable(Oid relationId)
+				Int64GetDatum(task->taskid);
+			nulls[Anum_pg_dist_background_tasks_depend_task_id - 1] = false;
+
+			values[Anum_pg_dist_background_tasks_depend_depends_on - 1] =
+				Int64GetDatum(dependingTaskIds[i]);
+			nulls[Anum_pg_dist_background_tasks_depend_depends_on - 1] = false;
+
+			HeapTuple newTuple = heap_form_tuple(
+				RelationGetDescr(pgDistbackgroundTasksDepend), values, nulls);
+			CatalogTupleInsert(pgDistbackgroundTasksDepend, newTuple);
+		}
+	}
+
+	if (pgDistbackgroundTasksDepend)
+	{
+		table_close(pgDistbackgroundTasksDepend, NoLock);
+	}
+	table_close(pgDistBackgroundTasks, NoLock);
+
+	return task;
+}
+
+
+/*
+ * ResetRunningBackgroundTasks finds all tasks currently in Running state and resets their
+ * state back to runnable.
+ *
+ * Tasks that were running during a failover, crash or restart will still be marked
+ * running after. This function assumes appropriate care by the caller that the task is in
+ * fact _not_ running at the moment on the primary coordinator of the formation.
+ *
+ * Any pid associated with the running tasks will be cleared back to the NULL value.
+ */
+void
+ResetRunningBackgroundTasks(void)
+{
+	const int scanKeyCount = 1;
+	ScanKeyData scanKey[1];
+	const bool indexOK = true;
+
+	Relation pgDistBackgroundTasks =
+		table_open(DistBackgroundTasksRelationId(), AccessShareLock);
+
+	/* pg_dist_background_tasks.status == 'running' */
+	ScanKeyInit(&scanKey[0], Anum_pg_dist_background_tasks_status,
+				BTEqualStrategyNumber, F_OIDEQ,
+				ObjectIdGetDatum(CitusTaskStatusRunningId()));
+
+	SysScanDesc scanDescriptor =
+		systable_beginscan(pgDistBackgroundTasks,
+						   DistBackgroundTasksStatusTaskIdIndexId(),
+						   indexOK, NULL, scanKeyCount,
+						   scanKey);
+
+	HeapTuple taskTuple = NULL;
+	while (HeapTupleIsValid(taskTuple = systable_getnext(scanDescriptor)))
+	{
+		Datum values[Natts_pg_dist_background_tasks] = { 0 };
+		bool isnull[Natts_pg_dist_background_tasks] = { 0 };
+		bool replace[Natts_pg_dist_background_tasks] = { 0 };
+
+		TupleDesc tupleDescriptor = RelationGetDescr(pgDistBackgroundTasks);
+		heap_deform_tuple(taskTuple, tupleDescriptor, values, isnull);
+
+		values[Anum_pg_dist_background_tasks_status - 1] =
+			ObjectIdGetDatum(CitusTaskStatusRunnableId());
+		isnull[Anum_pg_dist_background_tasks_status - 1] = false;
+		replace[Anum_pg_dist_background_tasks_status - 1] = true;
+
+		/* if there is a pid we need to signal the backend to stop */
+		if (!isnull[Anum_pg_dist_background_tasks_pid - 1])
+		{
+			int pid = DatumGetInt32(values[Anum_pg_dist_background_tasks_pid - 1]);
+			PGPROC *proc = BackendPidGetProc(pid);
+			const int sig = SIGTERM; /* we need to fully stop execution */
+			if (proc)
+			{
+				/* it is a postgres process managed by postmaster */
+#ifdef HAVE_SETSID
+				if (kill(-pid, sig))
+#else
+				if (kill(pid, sig))
+#endif
+				{
+					ereport(WARNING, (errmsg("could not send signal to process %d: %m",
+											 pid)));
+				}
+			}
```

Comment:
Should have a comment that this is heavily based on `pg_signal_backend`. Maybe even put this in a separate function.

Reply:
Guess that makes sense, since then I started using it elsewhere as well.